File: multiplication/src/test/java/learn/mssb/mssbmultiplication/challenge/ChallengeGeneratorServiceImplTest.java
package learn.mssb.mssbmultiplication.challenge;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Spy;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Random;

import static learn.mssb.mssbmultiplication.challenge.ChallengeGeneratorService.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.given;

@ExtendWith(MockitoExtension.class)
class ChallengeGeneratorServiceImplTest {

    private static final int BOUND_OF_RANDOM = MAXIMUM_FACTOR + 1 - MINIMUM_FACTOR;

    private ChallengeGeneratorService challengeGeneratorService;

    @Spy
    private Random random;

    @BeforeEach
    void setUp() {
        challengeGeneratorService = new ChallengeGeneratorServiceImpl(random);
    }

    @Test
    void randomChallenge() {
        final int randomForA = 20;
        final int randomForB = 30;
        given(random.nextInt(BOUND_OF_RANDOM)).willReturn(randomForA, randomForB);

        final Challenge challenge = challengeGeneratorService.randomChallenge();

        int expectedA = MINIMUM_FACTOR + randomForA;
        int expectedB = MINIMUM_FACTOR + randomForB;
        assertThat(challenge).isEqualTo(new Challenge(expectedA, expectedB));
    }
}

File: multiplicationkt/src/test/kotlin/challenge/ChallengeGeneratorServiceImplTest.kt
package challenge

import learn.mssbkt.multiplicationkt.challenge.Challenge
import learn.mssbkt.multiplicationkt.challenge.ChallengeGeneratorService
import learn.mssbkt.multiplicationkt.challenge.ChallengeGeneratorServiceImpl
import org.assertj.core.api.BDDAssertions.then
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.BDDMockito.given
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import kotlin.random.Random

@ExtendWith(MockitoExtension::class)
class ChallengeGeneratorServiceImplTest {

    private lateinit var challengeGeneratorService: ChallengeGeneratorService

    @Mock
    private lateinit var random: Random

    @BeforeEach
    fun setUp() {
        challengeGeneratorService = ChallengeGeneratorServiceImpl(random)
    }

    @Test
    fun `generated random factor is between expected limits`() {
        // given
        val maxToMinRange = 89
        given(random.nextInt(maxToMinRange)).willReturn(20, 30)
        // when
        val challenge = challengeGeneratorService.randomChallenge()
        // then
        then(challenge).isEqualTo(Challenge(factorA = 31, factorB = 41))
    }
}
